src/JustSaying/Messaging/Channels/IDownloadBuffer.cs

```diff
@@ -0,0 +1,109 @@
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using Amazon.SQS.Model;
+using JustSaying.AwsTools.MessageHandling;
+using JustSaying.Messaging.Monitoring;
+using Microsoft.Extensions.Logging;
+
+namespace JustSaying.Messaging.Channels
+{
+    internal interface IDownloadBuffer
+    {
+        Task Start(CancellationToken stoppingToken);
+        ChannelReader<IQueueMessageContext> Reader { get; }
+    }
+
+    internal class DownloadBuffer : IDownloadBuffer
+    {
+        private readonly Channel<IQueueMessageContext> _channel;
+        private readonly int _bufferLength;
+        private readonly ISqsQueue _sqsQueue;
+        private readonly ILogger _logger;
+
+        // todo: add logic around populating this
+        private readonly List<string> _requestMessageAttributeNames = new List<string>();
+
+        public ChannelReader<IQueueMessageContext> Reader => _channel.Reader;
+
+        public DownloadBuffer(
+            int bufferLength,
+            ISqsQueue sqsQueue,
+            ILoggerFactory logger)
+        {
+            _bufferLength = bufferLength;
+            _sqsQueue = sqsQueue;
+            _logger = logger.CreateLogger<IDownloadBuffer>();
+            _channel = Channel.CreateBounded<IQueueMessageContext>(bufferLength);
+        }
+
+        public async Task Start(CancellationToken stoppingToken)
+        {
+            ChannelWriter<IQueueMessageContext> writer = _channel.Writer;
+            try
+            {
+                while (!stoppingToken.IsCancellationRequested)
+                {
+                    var canWrite = await WaitToWriteAsync(writer, stoppingToken).ConfigureAwait(false);
+                    if (!canWrite) break;
+
+                    IList<Message> messages;
+                    using (_logger.TimedOperation("Receiving messages from SQS for queue {QueueName}", _sqsQueue.QueueName))
+                    {
+                        messages = await _sqsQueue
+                           .GetMessages(_bufferLength, _requestMessageAttributeNames, stoppingToken)
+                           .ConfigureAwait(false);
+
+                        if (messages == null) continue;
+                    }
+
+                    foreach (var message in messages)
```

Comment:
Not sure where it's worth addressing this so I'll do it here.

We're trying to take a slightly novel approach to cancellation here: we want to make it possible for users to trigger cancellation, and allow a grace period in which to process the remaining downloaded messages so that we can avoid the visibility timeout when a message is downloaded but not processed. 

In this PR we have things cancelling so that its top-down from the downloaders, everything else cancels based on its `Completion` task - nice. As an escape hatch for this behaviour if its undesired, we'd like to expose some configuration:
```csharp
WithBusCancellation(CancellationToken token)
WithBusCancellation(CancellationToken token, TimeSpan shutdownGracePeriod = 10sec)
```
that allows the user to pass in the lifetime cancellation token that would trigger immediate cancellation of all WIP.

So we haven't yet threaded tokens through everywhere yet, though we've got it on our list!

Reply:
Just an opinion here: if a message hasn't been passed to a handler, I'd let the visibility timeout do it's thing.

Reply:
I'm erring to Stu's point here, I was trying to do something similar in a different PR related to all this throttling stuff.